### PR TITLE
DAOS-9781 control: Parse bdev_list entries early

### PR DESCRIPTION
**Pull request by mjmac**
Instead of passing around strings which need to be parsed
and validated in different parts of the system, convert
the configuration to use structured device lists. This
will allow errors to be caught earlier and simplify
other parts of the codebase.

Signed-off-by: Michael MacDonald <mjmac.macdonald@intel.com>

### CODE REVIEW

File: src/control/server/storage/config.go [hunk 1]
```diff
@@ -259,6 +267,177 @@ func (sc *ScmConfig) Validate(class Class) error {
 	return nil
 }
 
+type stringSet map[string]struct{}
```
Comment (mjmac) [inline src/control/server/storage/config.go:270]: Oops, just noticed that this is a dupe of [common.StringSet](https://github.com/daos-stack/daos/blob/master/src/control/common/collection_utils.go#L20), added after I started this PR ages ago. I'll wait to see what kind of feedback I get before pushing an update.

File: src/control/server/storage/config.go [hunk 2]
```diff
@@ -259,6 +267,177 @@ func (sc *ScmConfig) Validate(class Class) error {
...
+	// As this is the most common use case, we'll make the embedded type's methods
+	// available directly on the type.
+	hardware.PCIAddressSet
+
+	// As a fallback for non-PCI bdevs, maintain a map of strings.
+	stringBdevSet stringSet
+}
+
+// maybePCI does a quick check to see if a string could possibly be a PCI address.
+func maybePCI(addr string) bool {
```
Comment (tanabarr) [inline src/control/server/storage/config.go:300]: why not use the hardware PCI methods to attempt to convert? this seems like duplication of those

File: src/control/server/storage/config.go [hunk 3]
```diff
@@ -259,6 +267,177 @@ func (sc *ScmConfig) Validate(class Class) error {
...
+	if bdl == nil {
+		return errors.New("nil BdevDeviceList")
+	}
+
+	if bdl.stringBdevSet == nil {
+		bdl.stringBdevSet = stringSet{}
+	}
+
+	for _, strAddr := range addrs {
+		if !maybePCI(strAddr) {
```
Comment (tanabarr) [inline src/control/server/storage/config.go:319]: here why not try to create a PCI address and if it fails then add to the string set?

**Review by tanabarr — DISMISSED** [review]

File: src/control/server/storage/config.go (hunk 3 above)
Comment (mjmac) [reply src/control/server/storage/config.go:319]: I wanted to be able to distinguish between strings that are clearly not PCI addresses (i.e. some other type of bdev) and malformed (mistyped, etc) PCI addresses in order to return the error early, here, instead of waiting until later to do the validation.

File: src/control/server/config/server_test.go [hunk 4]
```diff
@@ -629,7 +635,7 @@ func TestServerConfig_Parsing(t *testing.T) {
 								WithBdevDeviceList(MockPCIAddr(1), MockPCIAddr(1)),
 						))
 			},
-			expValidateErr: errors.New("bdev_list contains duplicate pci"),
+			expValidateErr: errors.New("valid PCI addresses"),
```
Comment (kjacque) [inline src/control/server/config/server_test.go:638]: I'm inclined to think the old message about duplicates is more useful

File: src/control/server/storage/config.go [hunk 5]
```diff
@@ -372,7 +549,7 @@ func (bc *BdevConfig) Validate(class Class) error {
 }
 
 // parsePCIBusRange takes a string of format <Begin-End> and returns the begin and end values.
-// Number base is detected from the string prefixes e.g. 0x for hexadecimal.
+// Number bdle is detected from the string prefixes e.g. 0x for hexadecimal.
```
Comment (kjacque) [inline src/control/server/storage/config.go:552]: Looks like an odd typo here

File: src/control/server/config/server_test.go [hunk 6]
```diff
@@ -41,6 +41,12 @@ var defConfigCmpOpts = []cmp.Option{
 		security.CertificateConfig{},
 	),
 	cmpopts.IgnoreFields(Server{}, "Path"),
+	cmp.Comparer(func(x, y *storage.BdevDeviceList) bool {
```
[the remark this answers is not in the document]
Comment (mjmac) [reply src/control/server/config/server_test.go:44]: I considered that, but I'm a little reluctant to clutter the exported surface with more test gunk. The mocks and other test-only exports already give me a bit of heartburn. I should take a look to see if there's some way we can prevent all of that from showing up in the godocs stuff; it wouldn't bother me as much then.

File: src/control/server/ctl_storage.go [hunk 7]
```diff
@@ -32,20 +32,21 @@ func (scs *StorageControlService) Setup() {
...
+	bdevAddrs := &storage.BdevDeviceList{}
 	for _, storageCfg := range scs.instanceStorage {
 		for _, tierCfg := range storageCfg.Tiers.BdevConfigs() {
 			if tierCfg.Class != storage.ClassNvme {
 				// don't scan if any tier is using emulated NVMe
 				return
 			}
-			cfgBdevs = append(cfgBdevs, tierCfg.Bdev.DeviceList...)
+			if err := bdevAddrs.Add(tierCfg.Bdev.DeviceList.Addresses()...); err != nil {
+				scs.log.Debugf("%s\n", errors.Wrap(err, "Failed to add bdev addresses"))
```
[the remark this answers is not in the document]
Comment (mjmac) [reply src/control/server/ctl_storage.go:43]: Nope, it's not. Copy-pasted that error from elsewhere. I've been slowly removing those stray newlines when I notice them, but didn't notice this one.

File: src/control/server/config/server_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (kjacque) [reply src/control/server/config/server_test.go:44]: That's fair. I'm less a fan of exporting the mocks, but a comparer for go-cmp seems potentially useful in a general way.

File: src/control/lib/control/auto_test.go [hunk 8]
```diff
@@ -998,6 +998,12 @@ func TestControl_AutoConfig_genConfig(t *testing.T) {
 					config.Server{},
 					sysfs.Provider{},
 				),
+				cmp.Comparer(func(x, y *storage.BdevDeviceList) bool {
+					if x == nil && y == nil {
+						return true
+					}
+					return x.Equals(y)
```
Comment (knard-intel) [inline src/control/lib/control/auto_test.go:1005]: Could be usefull to also test if x == nil || y == nil. Otherwise, the return statement could crash if x is nil and not y.

File: src/control/server/storage/bdev/backend_json_test.go [hunk 9]
```diff
@@ -74,7 +74,7 @@ func TestBackend_newSpdkConfig(t *testing.T) {
 		"config validation failure": {
 			class:          storage.ClassNvme,
 			devList:        []string{"not a pci address"},
-			expValidateErr: errors.New("unexpected pci address"),
+			expValidateErr: errors.New("valid PCI addresses"),
```
Comment (knard-intel) [inline src/control/server/storage/bdev/backend_json_test.go:77]: As suggested by @kjacque ,for a newcomer such as me, the old error message is more relevant.

File: src/control/server/storage/config.go [hunk 10]
```diff
@@ -259,6 +267,177 @@ func (sc *ScmConfig) Validate(class Class) error {
...
+	if len(bdl.stringBdevSet) > 0 && bdl.PCIAddressSet.Len() > 0 {
+		return errors.New("cannot mix PCI and non-PCI block device addresses")
+	}
+
+	return nil
+}
+
+func (bdl *BdevDeviceList) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var tmp []string
+	if err := unmarshal(&tmp); err != nil {
```
Comment (knard-intel) [inline src/control/server/storage/config.go:349]: Probably a newbie golang question, but do not see how it could work without giving at least one more input parameter as for the json Unmarshall function.

File: src/control/server/storage/config.go [hunk 11]
```diff
@@ -259,6 +267,177 @@ func (sc *ScmConfig) Validate(class Class) error {
...
+	var out []string
+	for s := range ss {
+		out = append(out, s)
+	}
+	sort.Strings(out)
+	return out
+}
+
+// BdevDeviceList represents a set of block device addresses.
+type BdevDeviceList struct {
```
Comment (knard-intel) [inline src/control/server/storage/config.go:290]: From what I understand, I see one drawbacks with the proposed way of dealing with non PCI address. If we are mixing PCI bdev and non PCI bdev, then the non PCI bdev will not be accessible. Maybe, we should not allow to mix the two kinds of Device when created with the fromStrings method.

File: src/control/server/config/server_test.go (hunk 4 above)
Comment (mjmac) [reply src/control/server/config/server_test.go:638]: The error changed here because we're no longer validating the bdev_list entries in `BdevConfig.Validate()` -- at this point we're just checking that if the config specifies nvme, then the list of block devices should not be empty. The test is now a little awkward because it's using the `.WithBdevDeviceList()` helpers to construct the config before testing it. As these helpers cannot return an error, what we get with duplicate devices added via the helper is an empty device list. I see this as an edge case test rather than something that will be commonly hit by end users.

To put it another way, if you put duplicate addresses in your server yaml's bdev_list, you'll still see an error like this: `bdev_list: duplicate PCI address 0000:81:00.0`.

File: src/control/lib/control/auto_test.go (hunk 8 above)
Comment (mjmac) [reply src/control/lib/control/auto_test.go:1005]: The implementation of [BdevDeviceList.Equals()](https://github.com/daos-stack/daos/pull/8032/files#diff-af190e716da9a8067f6406eb4291dc08a2afabe0de9cb23599f29651e5d98cf8R387) handles nil `x` or `y` already.

**Review by kjacque — APPROVED** [review]

**Review by tanabarr — APPROVED** [review]
